src/twisted/internet/defer.py

```diff
@@ -2052,16 +2053,21 @@ def _inlineCallbacks(
             # application code will have a tb_next, but it will *not* have a
             # second tb_next.
             assert appCodeTrace.tb_next is not None
-            if appCodeTrace.tb_next.tb_next:
+            assert appCodeTrace.tb_next.tb_next is not None
+            if appCodeTrace.tb_next.tb_next.tb_next:
                 # If returnValue was invoked non-local to the frame which it is
                 # exiting, identify the frame that ultimately invoked
                 # returnValue so that we can warn the user, as this behavior is
                 # confusing.
+                #
+                # Note that there's one additional level due to returnValue being
+                # deprecated
                 ultimateTrace = appCodeTrace
 
                 assert ultimateTrace is not None
                 assert ultimateTrace.tb_next is not None
-                while ultimateTrace.tb_next.tb_next:
+                assert ultimateTrace.tb_next.tb_next is not None
+                while ultimateTrace.tb_next.tb_next.tb_next:
```

Comment:
Maybe put the comment closer to where it's used
```suggestion
                # Note that there's one additional level due to returnValue being
                # deprecated
                assert ultimateTrace.tb_next.tb_next is not None
                while ultimateTrace.tb_next.tb_next.tb_next:
```

Reply:
Yeah, here we can be liberal with explanations as the code is quite complicated.